[CALCITE-3737][CALCITE-3780] Implement HOP and SESSION table functions

amaliujia opened the pull request:
see:  
https://issues.apache.org/jira/browse/CALCITE-3737
https://issues.apache.org/jira/browse/CALCITE-3780


Some highlights on this PR:
1. support HOP as a table function.
2. support SESSION as a table function.
3. rename "table-valued function" to "table function" to improve naming.

DonnyZone commented on core/src/main/java/org/apache/calcite/sql/SqlWindowTableFunction.java:
```diff
@@ -20,17 +20,16 @@
 
 import static org.apache.calcite.util.Static.RESOURCE;
 
+
```
extra line?

amaliujia replied:
thanks! I will hold the change a bit until receive enough comments (and will address all style like comments at a time).

amaliujia replied:
fixed!

DonnyZone commented on core/src/main/java/org/apache/calcite/sql/SqlWindowTableFunction.java:
```diff
@@ -44,28 +43,21 @@ public SqlWindowTableFunction(String name) {
-    return SqlOperandCountRanges.of(3);
-  }
-
-  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
```
Why we remove these methods? Each window table function has its own implementation?

amaliujia replied:
Yes. Each windowing table-valued function shall check their own operands. The check of DESCRIPTOR names are also left for each function (which makes sense as each function has its own interpretation of names).

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -796,4 +800,86 @@ static Expression windowSelector(
+        }
+        return takeOne();
+      }
+
```
extra line

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -744,9 +748,9 @@ static Expression generatePredicate(
   }
 
-  /** Generates a window selector which appends attribute of the window based on
+  /** Generates a window selector which appends attribute of the window bases on
```
will fix `bases` as a typo.

swathi858 commented:
Can we create another table function like tumble or hop??

amaliujia commented:
@swathi858 can you send an email to dev@calcite.apache.org for your questions. That would be an better place to answer your question.

34venu commented:
how to register table functions in apache calcite?

DonnyZone commented:
We'd better change the title of this PR accordingly. It is still "table-value". Or can we simplify it (e.g., "Implement HOP and SESSION table function for ...", and etc.)?

amaliujia commented:
Sounds good! Renamed the title to "Implement HOP and SESSION table functions", which should be easy to understand now.

DonnyZone commented on core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:
```diff
@@ -2273,8 +2275,14 @@ public void unparse(
-  /** TUMBLE as a table-value function. */
-  public static final SqlFunction TUMBLE_TVF = new SqlWindowTableFunction(SqlKind.TUMBLE.name());
+  /** TUMBLE as a table function. */
+  public static final SqlFunction TUMBLE_TVF = new SqlTumbleTableFunction();
```
Just of curious, is the "TVF" a commonly-used abbreviation?

DonnyZone commented on core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:
```diff
@@ -2312,7 +2320,7 @@ public void unparse(
   /** The {@code HOP} group function. */
   public static final SqlGroupedWindowFunction HOP =
-      new SqlGroupedWindowFunction(SqlKind.HOP.name(), SqlKind.HOP, null,
+      new SqlGroupedWindowFunction("$HOP", SqlKind.HOP, null,
```
Why we need to change the function name here?

amaliujia replied:
Here is the context about this change: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java#L2304

DonnyZone replied:
I see, thanks for explanation.

DonnyZone commented on core/src/main/java/org/apache/calcite/util/BuiltInMethod.java:
```diff
@@ -586,7 +587,11 @@
       "singleGroupResultSelector", Function1.class),
-  TUMBLING(EnumerableDefaults.class, "tumbling", Enumerable.class, Function1.class);
+  TUMBLING(EnumerableDefaults.class, "tumbling", Enumerable.class, Function1.class),
+  HOPPING(EnumUtils.class, "hopping", Enumerator.class, int.class, long.class,
```
Any special reasons to implement `tumbling` and `hopping`&`sessionize` in different places?

amaliujia replied:
I see. No special reason. Would you suggest I put it in the same class? If so, which class is better?

DonnyZone replied:
Current now, I personally suggest to put them in `EnumUtils`. It is a little strange in Linq4j. But if we will add more streaming functions later, I think we can create a separate class for them. It's more clear.

amaliujia replied:
Done!

DonnyZone commented on site/_docs/reference.md:
```diff
@@ -1871,14 +1872,44 @@ is named as "fixed windowing".
+will apply session with 20-minute inactive gap on rows from table orders. rowtime is the
+watermarked column of table orders that tells data completeness. Session is applied per product.
+
 ### Grouped window functions
```
Some comments:
1) Is it necessary to deprecate the manner of using them as grouped window function?
2) Update "Table-valued functions." to "Table functions".  
3) Could you add doc the [streaming](https://calcite.apache.org/docs/stream.html) page? I think we can do this work in this PR or after finishing the umbrella JIRA.

amaliujia replied:
Re: 1. Yes. First of all, Table function style streaming can cover what grouped window function provides. Then grouped window function has drawbacks: 1) magic *_start, * _end function to get window_start, window_end 2) enforce to us a GROUP BY (if users only want to do windowing streaming join without aggregation, they shouldn't be bothered to use a GROUP BY). 3) error prone syntax (why users cannot do GROUP BY TUMBLE(), HOP()? The syntax is correct for SQL).
So I think (and probably Julian also agreed) grouped window function can be deprecated. 

I will send an email to discuss the deprecation with community. So grouped window functions won't disappear silently. 


Re 2 and 3: nice suggestions. I will add a commit for the change.

amaliujia replied:
I have addressed 2.

Regarding 3, I would do it later. I would prefer to have a discussion about deprecating grouped window functions in mailing list first, which will give a much better idea how `streaming` page be updated (e.g. keep both style or just keep the new one)

amaliujia commented:
Thanks @DonnyZone for your review! I will address your comments this weekend!

Recently I am pretty busy in workdays.

amaliujia commented:
@DonnyZone have addressed comment and pushed the change.

I should be able to address future comments soon as my daily work has passed the intensive period.

amaliujia commented:
Thinking failed tests are not related to this PR: 

that is about `':elasticsearch:test'` or Pig related test.

DonnyZone commented:
Hi @amaliujia, thanks for your work! I will take another look in these couple of days.

amaliujia commented:
@DonnyZone 

just solved the merge conflict and re-triggered tests. Would you be able to take a look?

DonnyZone commented:
@amaliujia Sorry for the late reply. I will take a look ASAP.

DonnyZone commented on core/src/main/java/org/apache/calcite/sql/SqlSessionTableFunction.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
+                                             boolean throwOnFailure) {
```
nitpicking: indent

amaliujia replied:
oops. Fixed

DonnyZone commented:
This PR has two main commits (CALCITE-3737 and CALCITE-3780) and other minor commits.
For merge convenience, could you please squash other minor ones into its main commit?

amaliujia commented:
@DonnyZone squashed commits to two commits.

DonnyZone commented:
Thanks! There are some failed tests. Could you please make an investigation?

danny0405 commented:
I would try to give a final review this week, let's make this into 1.23 ~

amaliujia commented:
Thank you both! Sounds great to make it released in 1.23.0!


And I checked the failed tests. I don't think they are relevant to this PR. E.g. Tried one failed test and it passes locally.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -796,4 +803,244 @@ static Expression windowSelector(
+
+    SessionizationEnumerator(Enumerator<Object[]> inputEnumerator,
+        int indexOfWatermarkedColumn, int indexOfKeyColumn, long gap) {
+      this.inputEnumerator = inputEnumerator;
```
Give doc for these parameters. Same for other codes.

amaliujia replied:
Done.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -796,4 +803,244 @@ static Expression windowSelector(
+
+  /**
+   * Create enumerable implementation that applies sessionization to elements from the input
+   * enumerator based on a specified key. Elements are windowed into sessions separated by
```
`Create` -> `Creates`, same for other codes.

amaliujia replied:
Done.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -796,4 +803,244 @@ static Expression windowSelector(
+      // thus firstly should take the current element.
+      elements.add(inputEnumerator.current());
+      // sessionization needs to see all data.
+      while (inputEnumerator.moveNext()) {
```
Is this true ? Shouldn't we rely on the watermark and gap to decide if a session is expired ? What about the late arrive data ?

This implementation should not be blocking.

amaliujia replied:
It is true because this is batch only implementation.

For streaming cases, it needs to rely on watermark to know when it's ready to emit data. For batch cases, technically watermark always advances to the positive infinite at the beginning to mark all data has arrived (because it is bounded, or pre-known input data set). So this implementation is correct for batch case, because "watermark = positive infinity" means all data need to be processed.

I do have a plan to introduce watermark into calcite, and along with it, there should also be a test unbounded/streaming input source be built. By doing so we should be able to migrate current batch only implementation to batch-streaming unified implementation, which depends on watermark. However it will take some time to reach that point (with the support of community).

danny0405 replied:
I hear your concerns, but the operator did have a param named watermark.

amaliujia replied:
Sorry I might have missed something. What's the param that is named watermark? You mean the watermarked column, e.g. timestamp column specified by descriptor?

danny0405 replied:
Then why there are param named `indexOfWatermarkColumn` ?

amaliujia replied:
Watermark is a concept that is used to indicate the completeness of data, and it is not necessary an implementation.

The naming of "WatermarkedColumn" really means two things:
1. This is the column that provide event timestamp, which has the "event timestamp" semantic. (You can find "Event time vs. processing time" section in https://www.oreilly.com/radar/the-world-beyond-batch-streaming-101/ to know more about this semantic)
2. Because it has "event timestamp" semantic, we reply on "watermark" to know the data completeness and define late data. 

And windowing implemented by these series of PRs are coupled with "event timestamp" semantic, or we can just call it "event timestamp windowing" semantic.

What does it mean for batch and streaming sources? Say "hey,  this is the watermarked column, which gives the event timestamp. Let's apply windowing on this column. Wait a minute, how do I know the data is complete, thus I can close a window?"

For streaming source: "Hey it is streaming case, and we have watermark! If the watermark passes the end of window, you can mark the window is closed and all future data falls into that window is late".

For batch source: "Hey it is batch case, all data is known and there is no late data. Just close the window when all data are processed. And watermark also applies because when you ask the watermark, it always tell you that you can close the window". **In this case we can either implement watermark, which explicitly tell all window can be closed, or we just close them because this is batch. After all, it is just an implementation choice**, 

I think it is really a high-level argument and depends on if you accept this statement: batch is a special case of streaming. If you do accept this statement, using "watermark" is not wrong. Batch case can use all streaming terms (note that I am not saying using the same implementation). Also in the future, after streaming sources are built, either we change code here, or not, the behavior of what is build in this PR won't change cause it matches with "event timestamp windowing" semantic. So it is **backward compatible**.

danny0405 replied:
I do understand what a watermark is, no need to explain more, what i have concern upon is that the implementation has no relationship with a "watermark" definition but in the function doc, we represent it everywhere, which i think is not in consistent.

amaliujia replied:
Ok what's the recommendation then? Do you want "watermark" as a word removed?

danny0405 replied:
How about we add some notes in the function document that we only implements batch group window.

amaliujia replied:
Sure. Added a note for each windowing implementation to say they work for batch only.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:
```diff
@@ -796,4 +803,244 @@ static Expression windowSelector(
+        int indexOfWatermarkedColumn, long emitFrequency, long intervalSize) {
+      this.inputEnumerator = inputEnumerator;
+      this.indexOfWatermarkedColumn = indexOfWatermarkedColumn;
+      this.emitFrequency = emitFrequency;
```
`indexOfWatermarkedColumn ` -> `wmIndex` ?

amaliujia replied:
I prefer to keep "watermark". This is the relatively new concept for Calcite, so abbreviation could make people confused. 

Actually I have observed it on "TVF" (table-valued function or table function). Seems like people are already confused on such abbreviation.

danny0405 replied:
Use "table function" sounds better, i don't really know what a "table-valued function" means.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -966,8 +970,8 @@ public MatchImplementor get(final SqlMatchFunction function) {
-  public TableValuedFunctionCallImplementor get(final SqlWindowTableFunction operator) {
-    final Supplier<? extends TableValuedFunctionCallImplementor> supplier =
+  public TableFunctionCallImplementor get(final SqlWindowTableFunction operator) {
+    final Supplier<? extends TableFunctionCallImplementor> supplier =
```
Also remove the values keyword from the commit message.

amaliujia replied:
Done.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -3143,11 +3147,62 @@ private Expression normalize(SqlTypeName typeName, Expression e) {
+        Expression inputEnumerable, RexCall call, PhysType inputPhysType, PhysType outputPhysType) {
+      Expression intervalExpression = translator.translate(call.getOperands().get(2));
+      Expression intervalExpression2 = translator.translate(call.getOperands().get(3));
+      RexCall descriptor = (RexCall) call.getOperands().get(1);
```
Give these `intervalExpression ` more meaningful name. e.g. `windowSize` and `slidingInterval` ?

amaliujia replied:
Done.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:
```diff
@@ -2296,8 +2298,14 @@ public void unparse(
+
+  /** SESSION as a table function. */
+  public static final SqlFunction SESSION_TVF = new SqlSessionTableFunction();
 
```
What does the `TVF` mean ? How about just name it `HOP_FUN`

amaliujia replied:
TVF is the abbrev for "table-valued function".  To clarify, "table-valued function" is the term used in SQL standard. But seems like people in Calcite community like the short version: table function.

I can rename it, but it will be better to use, e.g. HOP_TABLE_FUNCTION. What do you think?

danny0405 replied:
I would prefer to name it directly "HOP" and rename the old one to "_HOP", after all, we already rename the group window with prefix "$".

amaliujia replied:
How about put this rename in a deprecation process? Directly renaming is not backward compatible. I plan to send a discussion email to community to discuss the deprecation of group window functions. We can decide how to rename it based on the discussion (e.g. if people don't want to deprecate group window functions, it probably means that they want keep `HOP` for hopping by group windowing function).

amaliujia replied:
My personal idea is to fully deprecate group window function. HOP,TUMBLE and SESSION will be reserved for table function windowing only.

danny0405 replied:
Go ahead and send the mail ~

amaliujia replied:
After a second thought, I think I will just make the change about "HOP" and "_HOP" for old and new in this PR, because as you have said, I have already used "$" in the name.


The email discussion I want is about a bigger discussion on if we should fully deprecate group window function. I will do it later and that discussion is not relevant to this PR.

amaliujia replied:
There is a naming constraint: 

` Name '_TUMBLE' must match pattern '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. `

So the realistic naming is `TUMBLE_`. Do you think it is a good idea to rename all group window function operator with "_" as suffix?

danny0405 replied:
I'm fine with that.

amaliujia replied:
Used , for example, `TUMBLE_OLD` as the naming for grouped window functions.

danny0405 commented on core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:
```diff
@@ -1778,15 +1778,16 @@ public final Sql sql(String sql) {
-  @Test void testTableValuedFunctionTumbleWithSubQueryParam() {
+  @Test public void testTableFunctionTumbleWithSubQueryParam() {
     final String sql = "select *\n"
         + "from table(tumble((select * from Shipments), descriptor(rowtime), INTERVAL '1' MINUTE))";
```
Also add tests for `SESSION` and `HOP`.

amaliujia replied:
Done

danny0405 commented on site/_docs/reference.md:
```diff
@@ -1875,14 +1876,44 @@ is named as "fixed windowing".
 |:-------------------- |:-----------
-| TUMBLE(table, DESCRIPTOR(column_name), interval [, time ]) | Indicates a tumbling window of *interval* for *datetime*, optionally aligned at *time*. Tumbling is applied on table in which there is a watermarked column specified by descriptor.
+| TUMBLE(table, DESCRIPTOR(datetime), interval [, time ]) | Indicates a tumbling window of *interval* for *datetime*, optionally aligned at *time*.
 
```
Do we really support `[, time ]` ?

amaliujia replied:
yes you are right. We don't support it so far.

Do you prefer me to support it in this PR, or a future PR (I can modify documentation accordingly)?

danny0405 replied:
Either is fine, we should keep our document correct and exact.

amaliujia replied:
Done!

amaliujia commented:
@danny0405 I think I have addressed actionable comments. Please take another look.

danny0405 commented on core/src/main/codegen/templates/Parser.jj:
```diff
@@ -6065,10 +6065,22 @@ SqlCall GroupByWindowingCall():
+        <SESSION> { s = span(); }
+        args = UnquantifiedFunctionParameterList(ExprContext.ACCEPT_SUB_QUERY) {
+            return SqlStdOperatorTable.SESSION.createCall(s.end(this), args);
+        }
```
We can combine these
```java
args = UnquantifiedFunctionParameterList(ExprContext.ACCEPT_SUB_QUERY) {
            return SqlStdOperatorTable.SESSION.createCall(s.end(this), args);
        }
```
Use another variable to keep the operator and move the common parse logic to the java code block.

amaliujia replied:
Done.

danny0405 replied:
`span()` can also be promoted.

amaliujia replied:
I tried to change to this way
```
(
        <TUMBLE>
        {
            op = SqlStdOperatorTable.TUMBLE_OLD;
        }
    |
        <HOP>
        {
            op = SqlStdOperatorTable.HOP_OLD;
        }
    |
        <SESSION>
        {
            op = SqlStdOperatorTable.SESSION_OLD;
        }
    )
    args = UnquantifiedFunctionParameterList(ExprContext.ACCEPT_SUB_QUERY) {
        s = span();
        return op.createCall(s.end(this), args);
    }
```
But the parser didn't work as expected.

Any thing I have missed?

danny0405 replied:
No worries, i would promote that when i merged the PR.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/SqlSessionTableFunction.java:
```diff
@@ -0,0 +1,71 @@
+
+  @Override public String getAllowedSignatures(String opNameToUse) {
+    return getName() + "(TABLE table_name, DESCRIPTOR(col), "
+        + "DESCRIPTOR(col), datetime interval)";
```
Why we must have 2 `DESCRIPTOR` here ?

amaliujia replied:
For session, usually common usage is per key. E.g. per user session. That's what the second descriptor used for: offer a session key.

The group window function does not have this problem because that style of windowing happens at GROUP BY, where users can put separate GROUP BY key as session key.

danny0405 replied:
I mean why couldn't we have one `DESCRIPTOR` with 2 columns ?

amaliujia replied:
Because descriptor only provide column names so far. If there is only one descriptor in which there are two column names, it will have a problem to know which one is the timestamp column that apply windowing upon and which one is the session key(note that session could be timestamp type as well). Of course a workaround is to define, say: the first column name is for windowing, and the second column is for session key. But this is error-prone, both from user usage and documentation: users have to use the right column name on the right position and documentation needs to explain the meaning of column position.

Two descriptors is better as each one has its clear semantic and require one column only. 

So I think it's a good API design than single descriptor.

danny0405 replied:
I don't think so, use two descriptors also has such concerns, people would also ask "which column is timestamp column"

amaliujia replied:
It's ok that you don't think that benefit holds. Even if so, then seems to me that there is no difference between one descriptor and two descriptors. What do you think?

danny0405 replied:
The `descriptor` keyword itself is too verbose, why we must have that ? Couldn't we just put a column name there directly ?

amaliujia replied:
In fact, it is a feature defined in SQL standard 2016: Polymorphic table functions(https://issues.apache.org/jira/browse/CALCITE-2270). `DESCRIPTOR` is designed to specify columns explicitly in table functions. That was one of the purpose  it was created for.

danny0405 replied:
Thanks for the explanations, personally i think one `DESCRIPTOR` is more concise, i'm wondering what other people think ?

danny0405 replied:
Can the session key be multiple keys ?

amaliujia replied:
This is a good point. In my implementation, it is single key only (cause that is the typical usage). But compound key is feasible and valid use case.

If there is a need to support compound key, then I think two descriptors will be better.

danny0405 replied:
Yes, two descriptor is more feasible to support multiple session keys, we better add a test in  SqlToRelConverterTest to support multiple session keys.

amaliujia replied:
Added a `testTableFunctionSessionCompoundSessionKey` in `SqlToRelConverterTest` for verifying that generated plan is correct when there are multiple session key.

amaliujia replied:
Another reason why we use descriptor than just a column name: https://issues.apache.org/jira/browse/CALCITE-3964

Basically there is a missing feature of `DESCRIPTOR`: an optional data type, which can be used to do type validation. It's useful for event timestamp windowing because such windowing should be apply on TIMESTAMP column. `DESCRIPTOR` will be handy to validate if user query applies windowing on a right column.

danny0405 replied:
From the paper, the DESCRIPTOR only appears in the function definition, that is, it is in the DDL, here we are not creating a function, but use it directly, i'm not sure if we should use the descriptor, as for the first argument, i would fix it as a cursor.

danny0405 commented:
I have fire a fix for the mis-leading first operand of `TUMBLE` `RexCall`, please review if you have time https://github.com/apache/calcite/pull/1943/files

amaliujia commented:
Thanks for the PR. Have LGTMed it. 

I will rebase and make changes in my PR accordingly after #1943 is merged.

amaliujia commented:
Rebased and resolved conflicts.

amaliujia commented:
@danny0405 

Since I am seeing Calcite 1.23.0 release is under discussion in dev@calcite, do you think if CALCITE-3955 should be a blocker for this PR?

My personal opinion is we can merge this PR while working on CALCITE-3955 separately (I am following up CALCITE-3955). Basic arguments are 1) TUMBLE is merged already. 2) This PR is still functional without CALCITE-3955

amaliujia commented:
@DonnyZone @danny0405 

seems like Calcite community has set 05/11/2020 as the cutoff date for 1.23.0 RC1. Could we reach a conclusion on if we can make this change in 1.23.0?

DonnyZone commented:
@amaliujia @danny0405 
Thanks for Danny's insightful review! I learned a lot from the conversation.
Current now, this PR LGTM. So I'm +1 on merging it into 1.23.0.

amaliujia commented:
Thank you very much!

A friendly reminder that per current schedule, 1.23.0 RC1 will be cut by next Monday.

DonnyZone commented:
@amaliujia I noticed Danny says he will conduct some work when merging the PR.

DonnyZone commented:
I'd like to merge the PRs first. This issue has been blocked several months.
Extra work can be done in following-up tickets.

amaliujia commented:
Both of your effort are much appreciated! Thank you!

I have logged https://issues.apache.org/jira/browse/CALCITE-3985 to track the work that Danny mentioned.
